core/services/functions/connector_handler.go

```diff
@@ -47,3 +79,105 @@ func (h *functionsConnectorHandler) Start(ctx context.Context) error {
 func (h *functionsConnectorHandler) Close() error {
 	return nil
 }
+
+func (h *functionsConnectorHandler) handleSecretsList(ctx context.Context, gatewayId string, body *api.MessageBody, fromAddr ethCommon.Address) {
+	type ListRow struct {
+		SlotID     uint   `json:"slot_id"`
+		Version    uint64 `json:"version"`
+		Expiration int64  `json:"expiration"`
+	}
+
+	type ListResponse struct {
+		Success      bool      `json:"success"`
+		ErrorMessage string    `json:"error_message,omitempty"`
+		Rows         []ListRow `json:"rows,omitempty"`
+	}
+
+	var response ListResponse
+	snapshot, err := h.storage.List(ctx, fromAddr)
+	if err == nil {
+		response.Success = true
+		response.Rows = make([]ListRow, len(snapshot))
+		for i, row := range snapshot {
+			response.Rows[i] = ListRow{
+				SlotID:     row.SlotId,
+				Version:    row.Version,
+				Expiration: row.Expiration,
+			}
+		}
+	} else {
+		response.ErrorMessage = fmt.Sprintf("Failed to list secrets: %v", err)
+	}
+
+	if err := h.sendResponse(ctx, gatewayId, body, response); err != nil {
+		h.lggr.Errorw("failed to send response to gateway", "id", gatewayId, "error", err)
+	}
+}
+
+func (h *functionsConnectorHandler) handleSecretsSet(ctx context.Context, gatewayId string, body *api.MessageBody, fromAddr ethCommon.Address) {
+	type SetRequest struct {
+		SlotID     uint   `json:"slot_id"`
+		Version    uint64 `json:"version"`
+		Expiration int64  `json:"expiration"`
+		Payload    []byte `json:"payload"`
+		Signature  []byte `json:"signature"`
+	}
+
+	type SetResponse struct {
+		Success      bool   `json:"success"`
+		ErrorMessage string `json:"error_message,omitempty"`
+	}
+
+	var request SetRequest
+	var response SetResponse
+	err := json.Unmarshal(body.Payload, &request)
+	if err == nil {
+		key := s4.Key{
+			Address: fromAddr,
+			SlotId:  request.SlotID,
+			Version: request.Version,
+		}
+		record := s4.Record{
+			Expiration: request.Expiration,
+			Payload:    request.Payload,
+		}
+		err = h.storage.Put(ctx, &key, &record, request.Signature)
+		if err == nil {
+			response.Success = true
+		} else {
+			response.ErrorMessage = fmt.Sprintf("Failed to set secret: %v", err)
+		}
+	} else {
+		response.ErrorMessage = fmt.Sprintf("Bad request to set secret: %v", err)
+	}
+
+	if err := h.sendResponse(ctx, gatewayId, body, response); err != nil {
+		h.lggr.Errorw("failed to send response to gateway", "id", gatewayId, "error", err)
+	}
+}
+
+func (h *functionsConnectorHandler) sendResponse(ctx context.Context, gatewayId string, requestBody *api.MessageBody, payload any) error {
+	payloadJson, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	msg := &api.Message{
+		Body: api.MessageBody{
+			MessageId: requestBody.MessageId,
+			DonId:     requestBody.DonId,
+			Method:    requestBody.Method,
+			Sender:    h.nodeAddress,
+			Payload:   payloadJson,
+		},
+	}
+	if err = msg.Sign(h.signerKey); err != nil {
```

Comment:
If we decide that Gateway should combine responses from multiple nodes then the signature night need to be included in the payload. We can discuss that separately.

Reply:
This sets `msg.Signature`, and it gives it back with `SendToGateway`. The rest depends on the connector/gateway logic, you can include it or not, right? I mean this is not the functions' connector_handler responsibility anymore.

Reply:
The problem is what to do if you want to aggregate responses from multiple nodes with signatures. You could pack multiple copies of Body into payload but that's wasteful. Another option is to add an extra signature inside payload and later combine those. We can discuss separately.

Reply:
My understanding: this is the connector, that executes as a part of a single node. I don't get why this would ever implement any aggregation logic. This is one side.
The other side is this: `SendToGateway` requires `api.Message` param, therefore I must populate all the Message fields. Otherwise this design looks inconsistent. I am considering all Gateway IO as a transport layer, which has independent signatures, etc. Transport is not meant to aggregate anything.
If I am wrong, then I don't get the whole design.

Reply:
Yes, the connector doesn't aggregate anything, I was talking about what happens next in the Gateway handler. Adding the signature here is correct, I was just flagging that we haven't implemented the aggregation part yet (in the Gateway handler) and it might be convenient to include another signature in the payload that covers a smaller set of fields. Again, let's discuss this separately.